src/Behat/Testwork/ServiceContainer/ExtensionManager.php

```diff
@@ -34,30 +32,60 @@
     private $debugInformation = array(
         'extensions_list' => array()
     );
+    /**
+     * @var ExtensionInstantiator[]
+     */
+    private $instantiators = array();
 
     /**
      * Initializes manager.
      *
-     * @param Extension[] $extensions     List of default extensions
-     * @param null|string $extensionsPath Base path where to search custom extension files
+     * @param Extension[]                         $extensions      List of default extensions
+     * @param ExtensionInstantiator[]|string|null $instantiators   Extension instantiators to use
      */
-    public function __construct(array $extensions, $extensionsPath = null)
+    public function __construct(array $extensions, $instantiators = null)
     {
         foreach ($extensions as $extension) {
             $this->extensions[$extension->getConfigKey()] = $extension;
         }
 
-        $this->extensionsPath = $extensionsPath;
+        // BC Layer, using default instantiators
+        // to be removed in 4.0
+        if (!is_array($instantiators)) {
+            @trigger_error(
+                'The $extensionsPath parameters is deprecated since 3.4 and will be changed into an array of'
+                . 'instantiators in 4.0. Use the constructor or the setExtensionsPath method on the'
+                . ' ExtensionFileInstantiator class instead.',
+                E_USER_DEPRECATED
+            );
+
+            $instantiators = array(new ExtensionClassInstantiator(), new ExtensionFileInstantiator($instantiators));
+        }
+
+        $this->instantiators = $instantiators;
     }
 
     /**
      * Sets path to directory in which manager will try to find extension files.
      *
      * @param null|string $path
+     * @deprecated since 3.4, to be removed in 4.0
+     *  (Set the extensions path through the proper instantiator instead)
      */
     public function setExtensionsPath($path)
     {
-        $this->extensionsPath = $path;
+        @trigger_error(
+            sprintf(
+                '%s is deprecated since 3.4, and will be removed in 4.0. Use the proper instantiator instead.',
+                E_USER_DEPRECATED
```

Comment:
this must not be inside the sprintf. Your code is broken. what should be inside the sprintf is `__FUNCTION__`.